File: reference-implementation/java/src/main/java/org/apexstandard/reference/TickEngine.java
package org.apexstandard.reference;

import java.time.Instant;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

/**
 * Market data simulation engine.
 * Runs a 2-second fixed-rate tick that performs a random walk on mid price,
 * aggregates candles, and detects wall-clock candle close boundaries.
 */
final class TickEngine {

    private static final double HALF_SPREAD = 0.0001;
    private static final double MAX_PIP_STEP = 0.0002;
    private static final int TICK_INTERVAL_MS = 2000;

    @FunctionalInterface
    interface QuoteCallback {
        void onQuoteUpdate(double mid, double bid, double ask);
    }

    @FunctionalInterface
    interface CandleCloseCallback {
        void onCandleClose(String timeframe, CandleState candle);
    }

    @FunctionalInterface
    interface CandleUpdateCallback {
        void onCandleUpdate(String timeframe);
    }

    @FunctionalInterface
    interface FeatureUpdateCallback {
        void onFeatureUpdate();
    }

    static class CandleState {
        double open;
        double high;
        double low;
        double close;
        int volume;
        String openTime;
        boolean complete;

        CandleState(double price) {
            this.open = price;
            this.high = price;
            this.low = price;
            this.close = price;
            this.volume = 0;
            this.openTime = Instant.now().toString();
            this.complete = false;
        }

        CandleState copy() {
            CandleState copy = new CandleState(open);
            copy.high = high;
            copy.low = low;
            copy.close = close;
            copy.volume = volume;
            copy.openTime = openTime;
            copy.complete = complete;
            return copy;
        }
    }

    private double mid = 1.0875;
    private final Map<String, CandleState> candles = new ConcurrentHashMap<>();
    private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor(r -> {
        Thread t = new Thread(r, "tick-engine");
        t.setDaemon(true);
        return t;
    });
    private ScheduledFuture<?> tickFuture;

    private QuoteCallback quoteCallback;
    private CandleCloseCallback candleCloseCallback;
    private CandleUpdateCallback candleUpdateCallback;
    private FeatureUpdateCallback featureUpdateCallback;

    TickEngine() {
        for (String tf : new String[]{"M1", "M5", "H1"}) {
            candles.put(tf, new CandleState(mid));
        }
    }

    void setQuoteCallback(QuoteCallback cb) {
        this.quoteCallback = cb;
    }

    void setCandleCloseCallback(CandleCloseCallback cb) {
        this.candleCloseCallback = cb;
    }

    void setCandleUpdateCallback(CandleUpdateCallback cb) {
        this.candleUpdateCallback = cb;
    }

    void setFeatureUpdateCallback(FeatureUpdateCallback cb) {
        this.featureUpdateCallback = cb;
    }

    synchronized void start() {
        if (tickFuture != null) return;
        tickFuture = scheduler.scheduleAtFixedRate(this::tick, 0, TICK_INTERVAL_MS, TimeUnit.MILLISECONDS);
    }

    synchronized void stop() {
        if (tickFuture != null) {
            tickFuture.cancel(false);
            tickFuture = null;
        }
    }

    void shutdown() {
        stop();
        scheduler.shutdownNow();
    }

    void forceCandleClose(String timeframe) {
        CandleState candle = candles.get(timeframe);
        if (candle == null) return;
        synchronized (candle) {
            candle.complete = true;
            CandleState snapshot = candle.copy();
            if (candleCloseCallback != null) {
                candleCloseCallback.onCandleClose(timeframe, snapshot);
            }
        }
        candles.put(timeframe, new CandleState(mid));
    }

    double getMid() {
        return mid;
    }

    private static double roundTo5(double value) {
        return Math.round(value * 100_000.0) / 100_000.0;
    }

    private void tick() {
        try {
            // Random walk
            double delta = (Math.random() - 0.5) * 2 * MAX_PIP_STEP;
            mid = roundTo5(mid + delta);

            double bid = roundTo5(mid - HALF_SPREAD);
            double ask = roundTo5(mid + HALF_SPREAD);

            // Update all candle states
            for (String tf : new String[]{"M1", "M5", "H1"}) {
                CandleState candle = candles.get(tf);
                if (candle != null) {
                    synchronized (candle) {
                        candle.high = Math.max(candle.high, mid);
                        candle.low = Math.min(candle.low, mid);
                        candle.close = mid;
                        candle.volume += 1;
                    }
                }
            }

            // Check wall-clock boundaries for candle closes
            java.time.ZonedDateTime now = java.time.ZonedDateTime.now(java.time.ZoneOffset.UTC);
            int seconds = now.getSecond();
            int minutes = now.getMinute();

            if (seconds < 2) {
                CandleState m1 = candles.get("M1");
                if (m1 != null && m1.volume > 0) {
                    forceCandleClose("M1");
                }
                if (minutes % 5 == 0) {
                    CandleState m5 = candles.get("M5");
                    if (m5 != null && m5.volume > 0) {
                        forceCandleClose("M5");
                    }
                }
                if (minutes == 0) {
                    CandleState h1 = candles.get("H1");
                    if (h1 != null && h1.volume > 0) {
                        forceCandleClose("H1");
                    }
                }
            }

            // Fire callbacks
            if (quoteCallback != null) {
                quoteCallback.onQuoteUpdate(mid, bid, ask);
            }
            if (candleUpdateCallback != null) {
                candleUpdateCallback.onCandleUpdate("M1");
            }
            if (featureUpdateCallback != null) {
                featureUpdateCallback.onFeatureUpdate();
            }
        } catch (Exception e) {
            System.err.println("Tick engine error: " + e.getMessage());
        }
    }
}

File: reference-implementation/rust/src/tick_engine.rs
use std::sync::{Arc, Mutex};

use tokio::sync::mpsc;
use tokio::task::JoinHandle;

use crate::helpers::now_iso;

const HALF_SPREAD: f64 = 0.0001;
const MAX_PIP_STEP: f64 = 0.0002;
const TICK_INTERVAL_MS: u64 = 2_000;
const HISTORY_LIMIT: usize = 300;

fn round_to_5(value: f64) -> f64 {
    (value * 100_000.0).round() / 100_000.0
}

#[derive(Debug, Clone)]
pub struct CandleState {
    pub open: f64,
    pub high: f64,
    pub low: f64,
    pub close: f64,
    pub volume: u64,
    #[allow(dead_code)]
    pub open_time: String,
    #[allow(dead_code)]
    pub complete: bool,
}

#[derive(Debug, Clone)]
pub enum TickEvent {
    QuoteUpdate {
        mid: f64,
        bid: f64,
        ask: f64,
    },
    CandleClose {
        timeframe: String,
        candle: CandleState,
    },
    CandleUpdate {
        timeframe: String,
    },
    FeatureUpdate,
}

struct TickEngineInner {
    mid: f64,
    price_history: Vec<f64>,
    candles: std::collections::HashMap<String, CandleState>,
}

impl TickEngineInner {
    fn new() -> Self {
        let mut candles = std::collections::HashMap::new();
        let mid = 1.0875;
        for tf in &["M1", "M5", "H1"] {
            candles.insert(tf.to_string(), fresh_candle(mid));
        }
        Self {
            mid,
            price_history: Vec::new(),
            candles,
        }
    }

    fn tick(&mut self) -> Vec<TickEvent> {
        let mut events = Vec::new();

        // Random walk
        let delta = (rand::random::<f64>() - 0.5) * 2.0 * MAX_PIP_STEP;
        self.mid = round_to_5(self.mid + delta);

        let bid = round_to_5(self.mid - HALF_SPREAD);
        let ask = round_to_5(self.mid + HALF_SPREAD);

        self.price_history.push(self.mid);
        if self.price_history.len() > HISTORY_LIMIT {
            self.price_history.remove(0);
        }

        // Update all candle states
        for tf in &["M1", "M5", "H1"] {
            if let Some(candle) = self.candles.get_mut(*tf) {
                if self.mid > candle.high {
                    candle.high = self.mid;
                }
                if self.mid < candle.low {
                    candle.low = self.mid;
                }
                candle.close = self.mid;
                candle.volume += 1;
            }
        }

        // Check wall-clock boundaries for candle closes
        let now = chrono::Utc::now();
        let seconds = now.format("%S").to_string().parse::<u32>().unwrap_or(0);
        let minutes = now.format("%M").to_string().parse::<u32>().unwrap_or(0);

        if seconds < 2 {
            if let Some(candle) = self.candles.get("M1") {
                if candle.volume > 0 {
                    events.extend(self.force_candle_close("M1"));
                }
            }

            if minutes % 5 == 0 {
                if let Some(candle) = self.candles.get("M5") {
                    if candle.volume > 0 {
                        events.extend(self.force_candle_close("M5"));
                    }
                }
            }

            if minutes == 0 {
                if let Some(candle) = self.candles.get("H1") {
                    if candle.volume > 0 {
                        events.extend(self.force_candle_close("H1"));
                    }
                }
            }
        }

        events.push(TickEvent::QuoteUpdate { mid: self.mid, bid, ask });
        events.push(TickEvent::CandleUpdate {
            timeframe: "M1".to_owned(),
        });
        events.push(TickEvent::FeatureUpdate);

        events
    }

    fn force_candle_close(&mut self, timeframe: &str) -> Vec<TickEvent> {
        let mut events = Vec::new();
        if let Some(candle) = self.candles.get_mut(timeframe) {
            candle.complete = true;
            let closed = candle.clone();
            events.push(TickEvent::CandleClose {
                timeframe: timeframe.to_owned(),
                candle: closed,
            });
            *candle = fresh_candle(self.mid);
        }
        events
    }
}

fn fresh_candle(price: f64) -> CandleState {
    CandleState {
        open: price,
        high: price,
        low: price,
        close: price,
        volume: 0,
        open_time: now_iso(),
        complete: false,
    }
}

pub struct TickEngine {
    inner: Arc<Mutex<TickEngineInner>>,
    handle: Option<JoinHandle<()>>,
    event_tx: mpsc::UnboundedSender<TickEvent>,
}

impl TickEngine {
    pub fn new(event_tx: mpsc::UnboundedSender<TickEvent>) -> Self {
        Self {
            inner: Arc::new(Mutex::new(TickEngineInner::new())),
            handle: None,
            event_tx,
        }
    }

    pub fn start(&mut self) {
        if self.handle.is_some() {
            return;
        }

        let inner = self.inner.clone();
        let tx = self.event_tx.clone();

        let handle = tokio::spawn(async move {
            let mut interval = tokio::time::interval(
                tokio::time::Duration::from_millis(TICK_INTERVAL_MS),
            );
            loop {
                interval.tick().await;
                let events = {
                    let mut engine = inner.lock().expect("tick engine mutex poisoned");
                    engine.tick()
                };
                for event in events {
                    if tx.send(event).is_err() {
                        return;
                    }
                }
            }
        });
        self.handle = Some(handle);
    }

    pub fn force_candle_close(&self, timeframe: &str) -> Option<CandleState> {
        let mut inner = self.inner.lock().expect("tick engine mutex poisoned");
        let events = inner.force_candle_close(timeframe);
        let tx = &self.event_tx;
        let mut closed_candle = None;
        for event in events {
            if let TickEvent::CandleClose { candle, .. } = &event {
                closed_candle = Some(candle.clone());
            }
            let _ = tx.send(event);
        }
        closed_candle
    }

    #[allow(dead_code)]
    pub fn stop(&mut self) {
        if let Some(handle) = self.handle.take() {
            handle.abort();
        }
    }
}
